SonarQube.Common/RoslynV1SarifFixer.cs

```diff
@@ -0,0 +1,107 @@
+// <copyright file="SarifUtilities.cs" company="SonarSource SA and Microsoft Corporation">
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using System.Diagnostics;
+
+namespace SonarQube.Common
+{
+    public static class RoslynV1SarifFixer
+    {
+        /// <summary>
+        /// Returns true if the given SARIF came from the VS 2015 RTM Roslyn, which does not provide correct output.
+        /// </summary>
+        public /* for test */ static bool IsSarifFromRoslynV1(string input)
+        {
+            // low risk of false positives / false negatives
+            return (input.Contains(@"""toolName"": ""Microsoft (R) Visual C# Compiler""")
+                && input.Contains(@"""productVersion"": ""1.0.0"""));
+        }
+
+        /// <summary>
+        /// Used to correct invalid SARIF generated by VS 2015 RTM Roslyn. 
+        /// Correctly escapes backslashes and quotes.
+        /// 
+        /// Returns true if the output is valid SARIF (even if there was no change).
+        /// 
+        /// Example invalid line:
+        /// "shortMessage": "message \test\ ["_"]",
+        /// 
+        /// Example valid line:
+        /// "shortMessage": "message \\test\\ [\"_\"]",
+        /// </summary>
+        public /* for test */ static bool FixRoslynV1Sarif(string input, out string output)
+        {
+            try {
+                JObject.Parse(input);
+
+                output = input; // no change necessary if input was valid
+                return true; 
+            } catch (JsonReaderException) // we expect invalid JSON
+            {
+                if (!IsSarifFromRoslynV1(input))
+                {
+                    // invalid SARIF that has not originated from VS 2015 RTM cannot be fixed
+                    output = input;
+                    return false;
+                } else 
+                {
+                    string[] inputLines = input.Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
+
+                    for (int i = 0; i < inputLines.Length; i++)
+                    {
+                        string line = inputLines[i];
+                        if (line.Contains(@"""uri"": ")
+                            || line.Contains(@"""shortMessage"": ")
+                            || line.Contains(@"""fullMessage"": ")
+                            || line.Contains(@"""title"": "))
+                        {                            
+                            line = line.Replace(@"\", @"\\");
+
+                            string[] subStrings = line.Split('"');
```

Comment:
This would be a better place to put the valid/invalid examples. Just an invalid example would do.